### DISCUSSION TITLE
Exposing a hiddenserver to a torhiddenclient in shadow-tor

### DISCUSSION THREAD

**Discussion by 1umpus** [shadow/shadow#678] [General]
Somewhat loosely related to [672](https://github.com/shadow/shadow/issues/672), but it's my understanding that the shadow-tor-plugin works in conjunction with the tgen plugin to generate traffic as well as route this traffic through Tor via the SOCKS proxy. For example, a hiddenserver and a torhiddenclient host can be declared in a shadow configuration XML file such that these two hosts run the tgen process plugin to simulate communication over Tor. I appreciate your advise if this line of thought is accurate or not.

In my case, I would like to send and receive data over TCP via these two hosts over shadow-tor. So far, I have created a hiddenserver and torhiddenclient host (as well as the other hosts presented in the provided shadow-plugin-tor/resource/shadowtor-minimal-config example) and I do make use of the shadow-plugin-extras/echo  example to make a TCP server on the hiddenserver and a TCP client on the torhiddenclient. Here's a [link](https://gist.github.com/1umpus/aa87b876d32adc32ba22c8d6c72217bd) to the example.xml file I created for the experiment. Additionally, I modified the hiddenserver's config torrc file (located at: tor.hiddenserver.torrc) prior to running the example so that the echo server's port is exposed as a Tor hidden service via: 

`HiddenServicePort 9999 127.0.0.1:9999`
----
Running shadow on the configuration file returns a successful status code of 0 and since I extended the echo plugin by calling `printf()` on the sent/recv'd buffers on both hosts, I am able to see that data has been successfully transmitted/received through shadow in the hosts/ output directory. 

However, I'm not sure that the TCP traffic between the two hosts is actually being routed through shadow-tor using my provided example. 

Do I need to modify the echo plugin to route TCP traffic through the SOCKS proxy (just like tgen is doing) in order for TCP traffic to be routed through shadow-tor? If so, how do you suggest the best way to accomplish this would be?

Has anyone had success in writing a simple echo TCP server in this way?

**Comment by robgjansen** [comment]
> I appreciate your advise if this line of thought is accurate or not.

Yes, your understanding seems correct to me :)

> However, I'm not sure that the TCP traffic between the two hosts is actually being routed through shadow-tor using my provided example.

Based on your configuration, your TCP echo client is connecting directly to your TCP echo server. If you wanted your traffic to run over Shadow-Tor, you would have to configure your client to the local tor process via the SOCKS port.

> Do I need to modify the echo plugin to route TCP traffic through the SOCKS proxy (just like tgen is doing) in order for TCP traffic to be routed through shadow-tor? If so, how do you suggest the best way to accomplish this would be?

Yes, your intuition here is correct. You need to connect to the SOCKS proxy that is running in Shadow-Tor. TGen does already implement the client-side of the SOCKS connection. You might find the TGen state machine where it steps through the SOCKS handshake useful:
https://github.com/shadow/tgen/blob/8cad828a8a1aefa97e2dc1b867a5dca3d6ae56b4/src/tgen-transport.c#L1035

You could use some of that same code. You could also implement the client side of the SOCK5 handshake yourself. It is pretty simple, and the protocol is outlined here:
https://en.wikipedia.org/wiki/SOCKS#SOCKS5

**Comment by 1umpus** [comment]
Thank you for helping me with this! I have gone ahead and implemented the SOCKS5 proxy client handshake and created a sample `tor-echo` plugin.

Here's a [link](https://github.com/breakerspace/shadow-plugin-extras/tree/master/tor-echo) for your reference. I will create a pull request if you would like to discuss merging the updated `shadow-plugin-extras` repo with the new tor plugin.

**Comment by robgjansen** [comment]
This is great! The blocking version that you wrote seems very simple and very easy to understand. I think others could find it useful and I would welcome a pull request to the shadow-plugin-extras repo. Thanks for filling out the readme so other people know how to use it :)

Please go ahead and file a pull request when you feel ready. In the meantime I'm going to mark this issue "closed".

**Comment by robgjansen** [comment]
(Sorry, just noticed you already file the pull request in shadow/shadow-plugin-extras#8)
